README.md
# worxLandroidS

Plugin Jeedom pour Robots tondeuses Worx Landroid modèles S/M/L

# Présentation
Ce plugin permet de se connecter aux tondeuses worx Landroid modèles S/M/L 

# Configuration du plugin

La connexion vers la tondeuse se fait à partir d'un serveur cloud en utilisant le compte utilisé lors de l'enregistrement de la tondeuse.

Les identifiants correspondent à ceux de l'application mobile. Vous devez attendre la fin de l'activation des dépendances pour permettre la communication avec la tondeuse par le protocole Mosquitto.

Une fois la sauvegarde effectuée, un nouvel équipement tondeuse va être créé automatiquement. En cas de soucis vous pouvez réactualiser les informations du cloud en utilisant la case à cocher correspondante et en effectuant une nouvelle sauvegarde.

L'arrêt du daemon permet de stopper la connexion avec la tondeuse. En cas du défaillance il se peut que le Daemon soit à l'arrêt et vous pouvez tenter un redémarrage.

# Utilisation

Le nom par défaut = LandroidS+adress Mac de la tondeuse

Le dashboard affiche:

-la date et heure de la dernière communication
-Etat de la tondeuse avec le code correspondant
-Description de l'erreur avec le code correspondant
-Le planning par jour avec l'heure de démarrage et d'arrêt
-La croix permet de mettre les horaires à 0
-L'autre bouton permet de récupérer les derniers horaires communiqués au plugin (10H - 17h par défaut)
-Edge signifie la coupe des bordures est planifié



docs/changelog.md
# Changelog

En cas d'absence de note dans ce chapitre, les mises à jour ne concernent que la doc, les traductions et des corrections mineures

## 27/06/2019
- correction date de fin de garantie (à la création de l'équipement)
- suppression code inutile 
- correction bug lié à virtualInfo

## 08/05/2019
- Ajout d'une info (virtualInfo) pour concatener plusieurs infos du plugin séparé par des virgules pour l'utilisation du widget worklandroid. Le widget est dispo sur le market des widgets. Recherche équipement pour récupérer l'info d'un autre équipement.
- remplacement des infos planning/xxxx/xxx par planning_xxxxx_xxxx suite à un changement du core jeedom

## 28/04/2019
- Diverses corrections
- Ajout de la fonction set_schedule pour modifier le planning de tonte d'un jour donné. Par défaut l'action n'est pas visible. Le but étant de faire de la planification à l'aide d'un scénario mais il est possible de rendre visible sur le widget si besoin.
- Format attendu: numéro jour;heure départ;durée en minutes;bordure
Exemples : 
  * 1;10:00;120;1 => lundi, démarrage à 10:00 pendant 120 minutes, coupe la bordure
  * 0;08:00;300;0 => dimanche, démarrage à 08:00 pendant 300 minutes, ne coupe pas la bordure


## 17/04/2019
- Nouvelle méthode d'authentification

## 03/04/2019
- Ajout coordonnées GPS si disponibles

## 29/03/2019
- Correction sur la fréquence de rafraîchissement des données

## 07/11/2018
- La nouvelle version du plugin nécessite la recréation des équipements, vous devez donc supprimer les équipements existants
- Gestion multi-tondeuses
- Détection automatique du type de tondeuse
- suppression mode retry

## 14/09/2018

- Correction de l'affichage sur mobile si le format du widget choisi est le "standard" jeedom. 
("Utiliser le widget préconfiguré" est décoché)

## 11/09/2018

- Ajout du paramètre type de tondeuses: Landroid version S / landroid version M (firmware 5.x)
(en cas de soucis vous pouvez cocher réinitialiser les paramètres dans la configuration du plugin et sauvegarder)
- Ajout de la fonction "pause"

## 09/07/2018

- Possibilité de définir son propre widget pour les commandes de type infos pour permettre l'affichage de données supplémentaires
- Modification des types d'info numériques (peut-être aussi fait manuellement ou en recréant l'équipement)

## 03/07/2018

- Ajout message création équipement 
- Message si version mosquitto non compatible
- correction bug affichage zone de départ

## 25/06/2018

- correction bug message sur cron30
- nouvelle tentative en cas après 15 secondes pour éviter certaines fausses alertes

## 16/06/2018

- modification du script d'installation pour tenter de résoudre les problème de version de mosquitto (version mini 1.4.1)
- Installation version mosquitto 1.5 si version mosquitto 1.3

## 14/06/2018

- Corrections des fonctions démarrer/arrêter.
- Modifications timeout si le serveur mosquitto n'envoie aucun message 
- changement du délai pluie manquant dans certains cas

## 09/06/2018

Ajout de nouvelles actions: 
- Ajout des délais de tonte après une pluie
- Ajout des actions off_today / on_today pour faciliter la gestion de l'activité du jour par scénarios (pour les jours fériés par exemple)

Autres modifications:
- Widget désormais modifiable (couleur/transparence...)
- Possibilité d'enlever certaines infos: errorCode, statusCode, totalDistance, batteryChargeCycle, rainDelay
- Affichage de la prochaine zone de tonte. C'est la zone de départ de la prochaine tonte ou de celle en cours. 
- Changement des infos en numérique pour permettre de faire des statistiques (évolution de la batterie par exemple)

## 06/06/2018

Modification des fréquences de mise à jour des infos: 
- Toutes les 2 minutes pendant la tonte
- Toutes les 30 minutes en dehors des periodes de tonte
- sur demande ou envoi de mise à jour du planning de fonctionnement.

## 04/06/2018

- Changement délai daemon et autres paramètres de connexion au serveur worx

## 01/06/2018

- Ajustement design widget
- remplacement id client mosquitto

## Mai 2018

Création du plugin



core/class/MQTT.class.php
<?php

/* This file is part of Jeedom.
*
* Jeedom is free software: you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
*
* Jeedom is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with Jeedom. If not, see <http://www.gnu.org/licenses/>.
*/
require_once dirname(__FILE__) . '/../../../../core/php/core.inc.php';

class MQTT extends eqLogic {
  public static function health() {
    $return = array();
    $socket = socket_create(AF_INET, SOCK_STREAM, 0);
    $server = socket_connect ($socket , config::byKey('mqttAdress', 'MQTT', '127.0.0.1'), config::byKey('mqttPort', 'MQTT', '1883'));
    $return[] = array(
      'test' => __('Mosquitto', __FILE__),
      'result' => ($server) ? __('OK', __FILE__) : __('NOK', __FILE__),
      'advice' => ($server) ? '' : __('Indique si Mosquitto est disponible', __FILE__),
      'state' => $server,
    );
    return $return;
  }

  public static function deamon_info() {
    $return = array();
    $return['log'] = '';
    $return['state'] = 'nok';
    $cron = cron::byClassAndFunction('MQTT', 'daemon');
    if (is_object($cron) && $cron->running()) {
      $return['state'] = 'ok';
    }
    $dependancy_info = self::dependancy_info();
    if ($dependancy_info['state'] == 'ok') {
      $return['launchable'] = 'ok';
    }
    return $return;
  }

  public static function deamon_start($_debug = false) {
    self::deamon_stop();
    $deamon_info = self::deamon_info();
    if ($deamon_info['launchable'] != 'ok') {
      throw new Exception(__('Veuillez vérifier la configuration', __FILE__));
    }
    $cron = cron::byClassAndFunction('MQTT', 'daemon');
    if (!is_object($cron)) {
      throw new Exception(__('Tache cron introuvable', __FILE__));
    }
    $cron->run();
  }

  public static function deamon_stop() {
    $cron = cron::byClassAndFunction('MQTT', 'daemon');
    if (!is_object($cron)) {
      throw new Exception(__('Tache cron introuvable', __FILE__));
    }
    $cron->halt();
  }

  public static function dependancy_info() {
    $return = array();
    $return['log'] = 'MQTT_dep';
    $return['state'] = 'nok';
    $cmd = "dpkg -l | grep mosquitto";
    exec($cmd, $output, $return_var);
    //lib PHP exist
    $libphp = extension_loaded('mosquitto');
    if ($output[0] != "" && $libphp) {
      $return['state'] = 'ok';
    }
    return $return;
  }

  public static function dependancy_install() {
    log::add('MQTT','info','Installation des dépéndances');
    $resource_path = realpath(dirname(__FILE__) . '/../../resources');
    passthru('sudo /bin/bash ' . $resource_path . '/install.sh ' . $resource_path . ' > ' . log::getPathToLog('MQTT_dep') . ' 2>&1 &');
    return true;
  }

  public static function daemon() {
    log::add('MQTT', 'info', 'Paramètres utilisés, Host : ' . config::byKey('mqttAdress', 'MQTT', '127.0.0.1') . ', Port : ' . config::byKey('mqttPort', 'MQTT', '1883') . ', ID : ' . config::byKey('mqttId', 'MQTT', 'Jeedom'));



/*[2018-05-05 21:44:33][INFO] : Connexion result :{"id":20055,"name":"Siegrist","email":"Sebastien.siegrist@gmail.com","created_at":"2017-07-03 16:30:52","updated_at":"2017-09-04 11:16:08","city":null,"address":null,"zipcode":null,"country_id":250,"phone":null,"birth_date":null,"gender":null,"newsletter_subscription":null,"user_type":"customer","active":true,"locale":"en","push_notifications":true,"location":null,"api_token":"uG7PB0xpI3JTv81ELaskDKdmAHHwPcS8","token_expiration":"2067-06-21 16:31:56","mqtt_client_id":"android-uuid\/v1","mqtt_endpoint":"a1optpg91s0ydf-2.iot.eu-west-1.amazonaws.com"}
*/

 //$client = new Mosquitto\Client('android-uuid/v1');




    $client = new Mosquitto\Client(config::byKey('mqttId', 'MQTT', 'Jeedom'));
    $client->onConnect('MQTT::connect');
    $client->onDisconnect('MQTT::disconnect');
    $client->onSubscribe('MQTT::subscribe');
    $client->onMessage('MQTT::message');
    $client->onLog('MQTT::logmq');
    $client->setWill('/jeedom', "Client died :-(", 1, 0);

    try {
      if (config::byKey('mqttUser', 'MQTT', 'none') != 'none') {
        $client->setCredentials(config::byKey('mqttUser', 'MQTT'), config::byKey('mqttPass', 'MQTT'));
      }
      $client->connect(config::byKey('mqttAdress', 'MQTT', '127.0.0.1'), config::byKey('mqttPort', 'MQTT', '1883'), 60);
//      $client->connect('a1optpg91s0ydf-2.iot.eu-west-1.amazonaws.com', '8883', 60);
   
        $client->subscribe(config::byKey('mqttTopic', 'MQTT', '#'), config::byKey('mqttQos', 'MQTT', 1)); // !auto: Subscribe to root topic
        log::add('MQTT', 'debug', 'Subscribe to topic ' . config::byKey('mqttTopic', 'MQTT', '#'));
      //$client->loopForever();
      while (true) { $client->loop(); }
    }
    catch (Exception $e){
      log::add('MQTT', 'error', $e->getMessage());
    }
  }

  public static function connect( $r, $message ) {
    log::add('MQTT', 'info', 'Connexion à Mosquitto avec code ' . $r . ' ' . $message);
    config::save('status', '1',  'MQTT');
  }

  public static function disconnect( $r ) {
    log::add('MQTT', 'debug', 'Déconnexion de Mosquitto avec code ' . $r);
    config::save('status', '0',  'MQTT');
  }

  public static function subscribe( ) {
    log::add('MQTT', 'debug', 'Subscribe to topics');
  }

  public static function logmq( $code, $str ) {
    if (strpos($str,'PINGREQ') === false && strpos($str,'PINGRESP') === false) {
      log::add('MQTT', 'debug', $code . ' : ' . $str);
    }
  }

  public static function message( $message ) {
    log::add('MQTT', 'debug', 'Message ' . $message->payload . ' sur ' . $message->topic);
    if (is_string($message->payload) && is_array(json_decode($message->payload, true)) && (json_last_error() == JSON_ERROR_NONE)) {
      //json message
      $nodeid = $message->topic;
      $value = $message->payload;
      $type = 'json';
      log::add('MQTT', 'info', 'Message json : ' . $value . ' pour information sur : ' . $nodeid);
    } else {
      $topicArray = explode("/", $message->topic);
      $cmdId = end($topicArray);
      $key = count($topicArray) - 1;
      unset($topicArray[$key]);
      $nodeid = implode($topicArray,'/');
      $value = $message->payload;
      $type = 'topic';
      log::add('MQTT', 'info', 'Message texte : ' . $value . ' pour information : ' . $cmdId . ' sur : ' . $nodeid);
    }

    $elogic = self::byLogicalId($nodeid, 'MQTT');
    if (!is_object($elogic)) {
      $elogic = new MQTT();
      $elogic->setEqType_name('MQTT');
      $elogic->setLogicalId($nodeid);
      $elogic->setName($nodeid);
      $elogic->setConfiguration('topic', $nodeid);
      $elogic->setConfiguration('type', $type);
      log::add('MQTT', 'info', 'Saving device ' . $nodeid);
      $elogic->save();
    }
    $elogic->setStatus('lastCommunication', date('Y-m-d H:i:s'));
    $elogic->save();

    if ($type == 'topic') {
    $cmdlogic = MQTTCmd::byEqLogicIdAndLogicalId($elogic->getId(),$cmdId);
    if (!is_object($cmdlogic)) {
      log::add('MQTT', 'info', 'Cmdlogic n existe pas, creation');
      $cmdlogic = new MQTTCmd();
      $cmdlogic->setEqLogic_id($elogic->getId());
      $cmdlogic->setEqType('MQTT');
      $cmdlogic->setSubType('string');
      $cmdlogic->setLogicalId($cmdId);
      $cmdlogic->setType('info');
      $cmdlogic->setName( $cmdId );
      $cmdlogic->setConfiguration('topic', $message->topic);
      $cmdlogic->save();
    }
    $elogic->checkAndUpdateCmd($cmdId,$value);

  } else {
      // payload is json
      $json = json_decode($value, true);
      foreach ($json as $cmdId => $value) {
        $topicjson = $nodeid . '{' . $cmdId . '}';
        log::add('MQTT', 'info', 'Message json : ' . $value . ' pour information : ' . $cmdId);
        $cmdlogic = MQTTCmd::byEqLogicIdAndLogicalId($elogic->getId(),$cmdId);
        if (!is_object($cmdlogic)) {
          log::add('MQTT', 'info', 'Cmdlogic n existe pas, creation');
          $cmdlogic = new MQTTCmd();
          $cmdlogic->setEqLogic_id($elogic->getId());
          $cmdlogic->setEqType('MQTT');
          $cmdlogic->setSubType('string');
          $cmdlogic->setLogicalId($cmdId);
          $cmdlogic->setType('info');
          $cmdlogic->setName( $cmdId );
          $cmdlogic->setConfiguration('topic', $topicjson);
          $cmdlogic->save();
        }
        $elogic->checkAndUpdateCmd($cmdId,$value);
      }
    }
  }

  public static function publishMosquitto($_id, $_subject, $_message, $_retain) {
    log::add('MQTT', 'debug', 'Envoi du message ' . $_message . ' vers ' . $_subject);
    $publish = new Mosquitto\Client(config::byKey('mqttId', 'MQTT', 'Jeedom') . '_pub_' . $_id);
    if (config::byKey('mqttUser', 'MQTT', 'none') != 'none') {
      $publish->setCredentials(config::byKey('mqttUser', 'MQTT'), config::byKey('mqttPass', 'MQTT'));
    }
    $publish->connect(config::byKey('mqttAdress', 'MQTT', '127.0.0.1'), config::byKey('mqttPort', 'MQTT', '1883'), 60);
    $publish->publish($_subject, $_message, config::byKey('mqttQos', 'MQTT', '1'), $_retain);
    for ($i = 0; $i < 100; $i++) {
      // Loop around to permit the library to do its work
      $publish->loop(1);
    }
    $publish->disconnect();
    unset($publish);
  }

}

class MQTTCmd extends cmd {
  public function execute($_options = null) {
    switch ($this->getType()) {
      case 'action' :
      $request = $this->getConfiguration('request','1');
      $topic = $this->getConfiguration('topic');
      switch ($this->getSubType()) {
        case 'slider':
        $request = str_replace('#slider#', $_options['slider'], $request);
        break;
        case 'color':
        $request = str_replace('#color#', $_options['color'], $request);
        break;
        case 'message':
        $request = str_replace('#title#', $_options['title'], $request);
        $request = str_replace('#message#', $_options['message'], $request);
        break;
      }
      $request = str_replace('\\', '', jeedom::evaluateExpression($request));
      $request = cmd::cmdToValue($request);
      MQTT::publishMosquitto($this->getId(), $topic, $request, $this->getConfiguration('retain','0'));
      }
      return true;
    }
  }
